pkg/receive/head_series_limiter.go

```diff
@@ -0,0 +1,180 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package receive
+
+import (
+	"context"
+	"net/http"
+	"net/url"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/thanos-io/thanos/pkg/errors"
+	"github.com/thanos-io/thanos/pkg/httpconfig"
+	"github.com/thanos-io/thanos/pkg/promclient"
+)
+
+// headSeriesLimit implements headSeriesLimiter interface.
+type headSeriesLimit struct {
+	mtx                    sync.RWMutex
+	limit                  map[string]uint64
+	tenantCurrentSeriesMap map[string]float64
+	defaultLimit           uint64
+
+	metaMonitoringURL    *url.URL
+	metaMonitoringClient *http.Client
+	metaMonitoringQuery  string
+
+	configuredTenantLimit *prometheus.GaugeVec
+	limitedRequests       *prometheus.CounterVec
+	metaMonitoringErr     prometheus.Counter
+
+	logger log.Logger
+}
+
+func NewHeadSeriesLimit(w WriteLimitsConfig, registerer prometheus.Registerer, logger log.Logger) *headSeriesLimit {
+	limit := &headSeriesLimit{
+		metaMonitoringURL:   w.GlobalLimits.metaMonitoringURL,
+		metaMonitoringQuery: w.GlobalLimits.MetaMonitoringLimitQuery,
+		defaultLimit:        w.DefaultLimits.HeadSeriesLimit,
+		configuredTenantLimit: promauto.With(registerer).NewGaugeVec(
+			prometheus.GaugeOpts{
+				Name: "thanos_receive_head_series_limit",
+				Help: "The configured limit for active (head) series of tenants.",
+			}, []string{"tenant"},
+		),
+		limitedRequests: promauto.With(registerer).NewCounterVec(
+			prometheus.CounterOpts{
+				Name: "thanos_receive_head_series_limited_requests_total",
+				Help: "The total number of remote write requests that have been dropped due to active series limiting.",
+			}, []string{"tenant"},
+		),
+		metaMonitoringErr: promauto.With(registerer).NewCounter(
+			prometheus.CounterOpts{
+				Name: "thanos_receive_metamonitoring_failed_queries_total",
+				Help: "The total number of meta-monitoring queries that failed while limiting.",
+			},
+		),
+		logger: log.With(logger, "component", "receive-head-series-limiter"),
+	}
+
+	// Record default limit with empty tenant label.
+	limit.configuredTenantLimit.WithLabelValues("").Set(float64(limit.defaultLimit))
+
+	// Initialize map for configured limits of each tenant.
+	limit.limit = map[string]uint64{}
```

Comment:
The `limit.limit` seems a bit confusing, would naming the parameter something along the lines `limitsPerTenant` make it clearer?